runtime/sema/gen/main.go

```diff
@@ -0,0 +1,952 @@
+		g.decls = append(g.decls, decl)
+	}
+}
+
+func (*generator) VisitVariableDeclaration(_ *ast.VariableDeclaration) struct{} {
+	panic("variable declarations are not supported")
+}
+
+func (g *generator) VisitFunctionDeclaration(decl *ast.FunctionDeclaration) (_ struct{}) {
+	if len(g.containerTypeNames) == 0 {
+		panic("global function declarations are not supported")
+	}
+
+	if decl.IsStatic() {
+		panic("static function declarations are not supported")
+	}
+
+	functionName := decl.Identifier.Identifier
+	fullTypeName := g.fullTypeName()
+	docString := g.declarationDocString(decl)
+
+	parameters := decl.ParameterList.Parameters
+
+	parameterTypeAnnotations := make([]*ast.TypeAnnotation, 0, len(parameters))
+	for _, parameter := range parameters {
+		parameterTypeAnnotations = append(
+			parameterTypeAnnotations,
+			parameter.TypeAnnotation,
+		)
+	}
+
+	g.addDecls(
+		goConstDecl(
+			functionNameVarName(fullTypeName, functionName),
+			goStringLit(functionName),
+		),
+		goVarDecl(
+			functionTypeVarName(fullTypeName, functionName),
+			typeExpr(&ast.FunctionType{
+				ReturnTypeAnnotation:     decl.ReturnTypeAnnotation,
+				ParameterTypeAnnotations: parameterTypeAnnotations,
+			}),
+		),
+		goConstDecl(
+			functionDocStringVarName(fullTypeName, functionName),
+			goRawLit(docString),
+		),
+	)
+
+	return
+}
+
+func (g *generator) declarationDocString(decl ast.Declaration) string {
+	identifier := decl.DeclarationIdentifier().Identifier
+	docString := strings.TrimSpace(decl.DeclarationDocString())
+
+	if len(docString) == 0 {
+		panic(fmt.Errorf(
+			"missing doc string for %s",
+			g.memberID(identifier),
+		))
+	}
+
+	// TODO: allow by splitting and wrapping in double quotes
+	if strings.ContainsRune(docString, '`') {
+		panic(fmt.Errorf("invalid ` in doc string for field %s", g.memberID(identifier)))
+	}
+
+	return trimLineSpaces(docString)
+}
+
+func (*generator) VisitSpecialFunctionDeclaration(_ *ast.SpecialFunctionDeclaration) struct{} {
+	panic("special function declarations are not supported")
+}
+
+func (g *generator) VisitCompositeDeclaration(decl *ast.CompositeDeclaration) (_ struct{}) {
+	var isResource bool
+
+	compositeKind := decl.CompositeKind
+	switch compositeKind {
+	case common.CompositeKindStructure:
+		break
+	case common.CompositeKindResource:
+		isResource = true
+	default:
+		panic(fmt.Sprintf("%s declarations are not supported", compositeKind.Name()))
+	}
+
+	typeName := decl.Identifier.Identifier
+
+	g.containerTypeNames = append(g.containerTypeNames, typeName)
+	defer func() {
+		g.containerTypeNames = g.containerTypeNames[:len(g.containerTypeNames)-1]
+	}()
+
+	var memberDeclarations []ast.Declaration
+
+	for _, memberDeclaration := range decl.Members.Declarations() {
+		ast.AcceptDeclaration[struct{}](memberDeclaration, g)
+
+		memberDeclarationKind := memberDeclaration.DeclarationKind()
+		switch memberDeclarationKind {
+		case common.DeclarationKindField,
+			common.DeclarationKindFunction:
```

Comment:
Is this validation needed? I guess members of structs/resources can only be functions or fields?
Others (e.g: pragma) are already errored at the visitor.

Reply:
Good point! Removed and added a comment in 0046c76ae160af2dbc93eb2ab60cc8ea92cb494d